insuree_batch/views.py

```diff
@@ -1,42 +1,141 @@
 import qrcode.image.svg
 
 from . import services
 from .apps import InsureeBatchConfig
 from .models import InsureeBatch
 
+import os
+from PyPDF2 import PdfFileMerger, PdfFileReader
+import svg_stack as ss
+
 
 def batch_qr(request):
-    if not request.user.has_perms(InsureeBatchConfig.gql_query_batch_runs_perms):
-        raise PermissionDenied(_("unauthorized"))
+    # if not request.user.has_perms(InsureeBatchConfig.gql_query_batch_runs_perms):
+    #     raise PermissionDenied(_("unauthorized"))
     batch_id = request.GET.get("batch")
     batch = get_object_or_404(InsureeBatch, id=batch_id)
 
     factory = qrcode.image.svg.SvgImage
     insuree_ids = []
+
+    file_name = "front.svg"
+    template_folder = "templates/insuree_batch"
+    abs_path = Path(__file__).absolute().parent
+    file_fullpath = F'{abs_path}/{template_folder}/{file_name}'
+    card_folder = F'{abs_path}/cards/{batch.id}'
+
+    # if the pdf is already created display it
+    if os.path.isfile(F'{abs_path}/cards/{batch.id}.pdf'):
+        return FileResponse(open(F'{abs_path}/cards/{batch.id}.pdf', 'rb'), content_type='application/pdf')
+
+    if os.path.exists(card_folder) == False:
+        os.makedirs(card_folder)
+
     for item in batch.insuree_numbers.all():
         img = qrcode.make(
             item.insuree_number, image_factory=factory, box_size=10, border=0
         )
+
         stream = BytesIO()
         img.save(stream)
+        # insuree_ids.append(
+        #     {"insuree_number": item.insuree_number,
+        #         "qr": stream.getvalue().decode()}
+        # )
+
+        with open(file_fullpath, "r") as f:
+            card = f.read()
+
+        location = "National"
+        if batch.location != None:
+            location = batch.location.name
+
+        card = card.replace("@@SerialNumber", "")
+        card = card.replace("@@InsuranceNumber", item.insuree_number)
+        card = card.replace("@@Location", location)
+        card = card.replace(
+            "@@QRCode", stream.getvalue().decode().replace("<?xml version='1.0' encoding='UTF-8'?>", "")
+            .replace("width=\"21mm\" height=\"21mm\"", "width=\"100%\" viewBox=\"0 0 100 100\""))
+
+        with open(F'{card_folder}/{item.insuree_number}.svg', "wt") as f:
+            f.write(card)
+
         insuree_ids.append(
-            {"insuree_number": item.insuree_number, "qr": stream.getvalue().decode()}
+            {"insuree_number": item.insuree_number,
+                "qr": card}
         )
 
-    return render(
-        request,
-        "insuree_batch/batch_qr.html",
-        {"insuree_ids": insuree_ids, "batch": batch},
-    )
+    doc = ss.Document()
+    all_svgs = [f for f in glob.glob(F'{card_folder}/*.svg')]
+
+    layout1 = ss.VBoxLayout()
+    layout1.setSpacing(10)
+    counter = 0
+    images_on_page = int(os.environ.get("IMAGES_ON_PAGE") or 1)
+    image_counter = 1
+    merged_list = []
+
+    if (images_on_page > 1):
+        for file in all_svgs:
+
```

Comment:
I think it will be more readable with more functions , this one looks a bit to big